### PR TITLE
util: Implement modulo for arith_uint256

### PR DESCRIPTION
**Pull request by nkostoulas**
I've implemented this for some side work and thought it might be of interest.

### CODE REVIEW

File: src/arith_uint256.cpp [hunk 1]
```diff
@@ -107,6 +107,31 @@ base_uint<BITS>& base_uint<BITS>::operator/=(const base_uint& b)
...
 }
 
+template <unsigned int BITS>
+base_uint<BITS>& base_uint<BITS>::operator%=(const base_uint& b)
+{
+    base_uint<BITS> div = b;     // make a copy, so we can shift.
+    base_uint<BITS> num = *this; // make a copy, so we can subtract.
+    int num_bits = num.bits();
+    int div_bits = div.bits();
+    if (div_bits == 0)
```
Comment (qmma70) [inline src/arith_uint256.cpp]: Could probably throw *before* making the copies?

File: src/arith_uint256.cpp [hunk 2]
```diff
@@ -107,6 +107,31 @@ base_uint<BITS>& base_uint<BITS>::operator/=(const base_uint& b)
...
+    if (div_bits > num_bits) // remainder is the same number
+        return *this;
+    int shift = num_bits - div_bits;
+    div <<= shift; // shift so that div and num align.
+    while (shift >= 0) {
+        if (num >= div) {
+            num -= div;
+        }
+        div >>= 1; // shift back.
+        shift--;
```
Comment (qmma70) [inline src/arith_uint256.cpp]: Minor: `--shift;` would be a better.

**Review by qmma70 — COMMENTED** [review]
How is this going to be used?

**Comment by sipa** [comment]
What is this useful for? AFAIK we don't use or need moduli with big numbers anywhere.

**Comment by laanwj** [comment]
Thanks for your contribution.

However, if there's no use case of this within bitcoin, it doesn't make sense for it to be merged here. `arith_uint256` exists for some specific uses in our code and is by no means a general bignum implementation. It's also not exported as a library for others.

**Pull request closed without merging**